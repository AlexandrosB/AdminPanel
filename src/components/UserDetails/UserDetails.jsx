import React, { Component } from 'react';
import PropTypes from 'prop-types';

require('./Static/userDetails.scss');

class UserDetails extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        email: PropTypes.string.isRequired,
        phone: PropTypes.string.isRequired,
        address: PropTypes.string.isRequired,
        company: PropTypes.string.isRequired,
        updateUserDetails: PropTypes.func.isRequired
    };

    constructor(props) {
        super(props);
        const {
            name, email, phone, address, company
        } = this.props;
        this.state = {
            name,
            email,
            phone,
            address,
            company
        };
    }

    componentDidUpdate(oldProps) {
        const newProps = this.props;
        if (oldProps.id !== newProps.id) {
            this.setState({
                name: newProps.name,
                email: newProps.email,
                phone: newProps.phone,
                address: newProps.address,
                company: newProps.company
            });
        }
    }

    /**
     * Change the value of
     * the selected field
     * @param event
     * @param type
     */
    updateFields = (event, type) => {
        this.setState({
            [type]: event.target.value
        });
    };

    /**
     * Cancel all the changes
     * and restore the original
     * values
     */
    cancelChanges = () => {
        const {
            name, email, phone, address, company
        } = this.props;
        this.setState({
            name,
            email,
            phone,
            address,
            company
        });
    };

    /**
     * Update user info with
     * the new values
     */
    updateUserDetails = () => {
        const {
                name, email, phone, address, company
            } = this.state,
            { updateUserDetails, id, photo } = this.props,
            data = {
                id,
                photo,
                name,
                company,
                email,
                phone,
                address
            };
        updateUserDetails(data);
    };

    render() {
        const {
                name, email, phone, address, company
            } = this.state,
            {
                name: oldName, email: oldEmail, phone: oldPhone, address: oldAddress, company: oldCompany
            } = this.props,
            enableButton = (name !== oldName || email !== oldEmail
                || phone !== oldPhone || address !== oldAddress
                || company !== oldCompany);
        return (
            <section className='userDetailsContainer'>
                <div className='formGroup'>
                    <div className='formElement'>
                        <span className='label'>Name</span>
                        <input
                            type='text'
                            value={ name }
                            placeholder='Enter name'
                            onChange={ (e) => this.updateFields(e, 'name') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Email address</span>
                        <input
                            type='text'
                            value={ email }
                            placeholder='Enter email address'
                            onChange={ (e) => this.updateFields(e, 'email') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Phone</span>
                        <input
                            type='text'
                            value={ phone }
                            placeholder='Enter phone'
                            onChange={ (e) => this.updateFields(e, 'phone') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Address</span>
                        <input
                            type='text'
                            value={ address }
                            placeholder='Enter address'
                            onChange={ (e) => this.updateFields(e, 'address') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Company</span>
                        <input
                            type='text'
                            value={ company }
                            placeholder='Enter company'
                            onChange={ (e) => this.updateFields(e, 'company') }
                        />
                    </div>
                    <section className='buttons'>
                        <button
                            className='cancel'
                            onClick={ this.cancelChanges }
                            disabled={ !enableButton }
                        >
                            <p>Cancel</p>
                        </button>
                        <button
                            className='update'
                            onClick={ this.updateUserDetails }
                            disabled={ !enableButton }
                        >
                            <p>Save</p>
                        </button>
                    </section>
                </div>
            </section>
        );
    }
}

export default UserDetails;
