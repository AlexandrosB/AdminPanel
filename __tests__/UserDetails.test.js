import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { USERS } from '../__testData__/testData';

import UserDetails from '../src/components/UserDetails/UserDetails';

describe('<UserDetails />', () => {
    const mockFnCb = jest.fn();
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<UserDetails
            id={ USERS[0].id }
            photo={ USERS[0].photo }
            name={ USERS[0].name }
            email={ USERS[0].email }
            phone={ USERS[0].phone }
            address={ USERS[0].address }
            company={ USERS[0].company }
            updateUserDetails={ mockFnCb }
        />);
    });

    it('should match the snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        wrapper;
    });

    it('renders a section that contains child components', () => {
        expect(wrapper.find('section.userDetailsContainer').exists()).toBe(true);
    });

    it('check the type of props', () => {
        const testData = USERS[0],
            props = {
                id: testData.id,
                photo: testData.photo,
                name: testData.name,
                email: testData.email,
                phone: testData.phone,
                address: testData.address,
                company: testData.company,
            },
            mountWrap = mount(<UserDetails { ...props } />);
        expect(mountWrap.prop('id')).toBeString();
        expect(mountWrap.prop('photo')).toBeString();
        expect(mountWrap.prop('name')).toBeString();
        expect(mountWrap.prop('email')).toBeString();
        expect(mountWrap.prop('phone')).toBeString();
        expect(mountWrap.prop('address')).toBeString();
        expect(mountWrap.prop('company')).toBeString();
    });
});
