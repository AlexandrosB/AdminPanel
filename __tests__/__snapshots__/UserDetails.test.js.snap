// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDetails /> should match the snapshot 1`] = `
<section
  className="userDetailsContainer"
>
  <div
    className="formGroup"
  >
    <div
      className="formElement"
    >
      <span
        className="label"
      >
        Name
      </span>
      <input
        onChange={[Function]}
        placeholder="Enter name"
        type="text"
        value="Bates Washington"
      />
    </div>
    <div
      className="formElement"
    >
      <span
        className="label"
      >
        Email address
      </span>
      <input
        onChange={[Function]}
        placeholder="Enter email address"
        type="text"
        value="bates.washington@zolarex.io"
      />
    </div>
    <div
      className="formElement"
    >
      <span
        className="label"
      >
        Phone
      </span>
      <input
        onChange={[Function]}
        placeholder="Enter phone"
        type="text"
        value="+1 (915) 447-2207"
      />
    </div>
    <div
      className="formElement"
    >
      <span
        className="label"
      >
        Address
      </span>
      <input
        onChange={[Function]}
        placeholder="Enter address"
        type="text"
        value="958 Brevoort Place, Ona, Maine, 2433"
      />
    </div>
    <div
      className="formElement"
    >
      <span
        className="label"
      >
        Company
      </span>
      <input
        onChange={[Function]}
        placeholder="Enter company"
        type="text"
        value="ZOLAREX"
      />
    </div>
    <section
      className="buttons"
    >
      <button
        className="cancel"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        <p>
          Cancel
        </p>
      </button>
      <button
        className="update"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        <p>
          Save
        </p>
      </button>
    </section>
  </div>
</section>
`;
